application-analytics-ui/src/main/resources/Analytics/Code/Macros/ExitPagesJSON.xml

```diff
@@ -0,0 +1,103 @@
+<?xml version="1.1" encoding="UTF-8"?>
+
+<!--
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+-->
+
+<xwikidoc version="1.4" reference="Analytics.Code.Macros.ExitPagesJSON" locale="">
+  <web>Analytics.Code.Macros</web>
+  <name>ExitPagesJSON</name>
+  <language/>
+  <defaultLanguage/>
+  <translation>0</translation>
+  <creator>xwiki:XWiki.Admin</creator>
+  <parent>WebHome</parent>
+  <author>xwiki:XWiki.Admin</author>
+  <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
+  <version>1.1</version>
+  <title>ExitPagesJSON</title>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>{{include reference='Analytics.Code.Macros.VelocityMacros'/}}
+
+{{velocity}}
+#if ($xcontext.action == 'get')
+  #analytics_livedataPagination()
+  ## Mapping between the livedata columns and the Matomo fields.
+  #set ($matomoKeysMapping = {
+    'pageTitle' : 'label',
+    'visits': 'nb_visits',
+    'exitRate' : 'exit_rate',
+    'exitNbVisits':'exit_nb_visits'
+  })
+  #set ($sort = 'nb_hits')
+  #set ($order = 'desc')
+  #if ($request.sort)
+    #set ($sort = $matomoKeysMapping.get($request.sort))
+    #set ($order = $request.dir)
+  #end
+
+  #analytics_prepareFilters($matomoKeysMapping)
+
+  #set ($parameters = {
+    'period' : $request.period,
+    'date' : $request.date,
+    'module' : 'API',
+    'method' : 'Actions.getPageUrls',
+    'format' : 'json',
+    'filter_limit' : $request.limitEntries,
+    'filter_sort_column' : $sort,
+    'filter_sort_order' : $order,
+    'expanded' : '1',
+    'flat' : '1'
+  })
+
+  ## Get the JSON response from Matomo.
+  #set ($analyticsResult = $services.analytics.makeRequest($parameters, $filters, 'MostViewedPages'))
+  #if ($offset &lt; $analyticsResult.size())
+    #set ($toIndex = $mathtool.min($mathtool.add($offset, $limit), $analyticsResult.size()))
+    #subSet($analyticsResult, $offset, $toIndex, $resultList)
+  #end
```

Comment:
Explain why you do the pagination yourself instead of passing the offset and limit to Matomo.

Reply:
It was an oversight on my part. I can directly use Matomo to retrieve columns from a specific offset. To do this, I need to add two new parameters to all requests: 'filter_limit' (which determines the number of entries I want) and 'filter_offset' (which specifies the starting index). Additionally, I must introduce two more parameters to ensure that Matomo can accurately locate instances when searching for a specific value in a field: 'filter_pattern' and 'filter_column', to which the pattern will be applied. I chose to take a subset because I couldn't figure out how to sort a column by a specific keyword and calculate the offset when I implemented this in the summer. I will make the changes for this PR and will create a new issue to apply the new offset method in other macros too.